### PR TITLE
add InteractiveSession and SessionManager

### PR DESCRIPTION
**Pull request by penghuo**
### Description
1. add InteractiveSession and SessionManager
 
### Issues Resolved
Phase 3. https://github.com/opensearch-project/sql/issues/2089
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+package org.opensearch.sql.spark.execution.session;
+
+import lombok.Data;
+import org.opensearch.sql.spark.client.StartJobRequest;
+
+@Data
+public class CreateSessionRequest {
+  private final StartJobRequest startJobRequest;
+  private final String datasourceName;
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java:14]: Do we require datasourceName in asyncQuery Request model?

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionState.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Getter;
+
+@Getter
+public enum SessionState {
+  NOT_STARTED("not_started"),
+  RUNNING("running"),
+  DEAD("dead"),
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionState.java:17]: Whats the difference between DEAD and FAILED?

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionModel.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.core.xcontent.XContentParserUtils;
+import org.opensearch.index.seqno.SequenceNumbers;
+
+/** Session data in flint.ql.sessions index. */
+@Data
+@Builder
+public class SessionModel implements ToXContentObject {
+  public static final String VERSION = "version";
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionModel.java:25]: What is version here?

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java (hunk 1 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java:14]: datasourceName is write to Session doc in flint.ql.sessions index. it is required for user perform doc level access control.

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionState.java (hunk 2 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionState.java:17]: * CLOSE_SESSION event transit state to DEAD.
* SPARK_APP FINISH SUCCESS event transit state to DEAD.

* SPARK_APP FINISH FAILED event transit state to FAILED

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionManager.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  public Optional<Session> getSession(SessionId sid) {
+    Optional<SessionModel> model = stateStore.get(sid);
+    if (model.isPresent()) {
+      InteractiveSession session =
+          InteractiveSession.builder()
+              .sessionId(sid)
+              .sessionStateStore(stateStore)
+              .serverlessClient(emrServerlessClient)
+              .sessionModel(model.get())
+              .build();
```
Comment (kaituo) [inline spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionManager.java:46]: Is it ok that the session's createSessionRequest is not set?

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionModel.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        .sessionType(INTERACTIVE)
+        .sessionId(sid)
+        .sessionState(NOT_STARTED)
+        .datasourceName(datasourceName)
+        .applicationId(applicationId)
+        .jobId(jobId)
+        .error(UNKNOWN)
+        .lastUpdateTime(System.currentTimeMillis())
+        .seqNo(SequenceNumbers.UNASSIGNED_SEQ_NO)
+        .primaryTerm(SequenceNumbers.UNASSIGNED_PRIMARY_TERM)
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionModel.java:140]: so these unassigned numbers can ensure version conflict happen when create session in the beginning?

Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionModel.java:140]: yes.

**Review by dai-chen — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionManager.java (hunk 4 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/SessionManager.java:45]: done. make sense

createSessionRequest  is open() method parameters.

**Review by dai-chen — APPROVED** [review]

**Review by kaituo — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java (hunk 1 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java:14]: I am actually asking for this api _plugins/_async_query. I think we need it. 
Should we perform datasource authorization in both the calls?

**Review by vamsi-amazon — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java (hunk 1 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/execution/session/CreateSessionRequest.java:14]: 1. for create session, _plugins/_async_query/sessions, require datasource, address it in separate PR. 
2. we need datasource authZ for both _plugins/_async_query/sessions and _plugins/_async_query API.

**Comment by penghuo** [comment]
test passed
bwc test failed, sql-2.12 not ready yet.

**Pull request merged**